dexcom/data_range.go

```diff
@@ -0,0 +1,143 @@
+package dexcom
+
+import (
+	"time"
+
+	"github.com/tidepool-org/platform/structure"
+	structureValidator "github.com/tidepool-org/platform/structure/validator"
+)
+
+type DataRange struct {
+	Start *DateRange `json:"start,omitempty"`
+	End   *DateRange `json:"end,omitempty"`
+}
+
+type DateRange struct {
+	SystemTime  *Time `json:"systemTime,omitempty"`
+	DisplayTime *Time `json:"displayTime,omitempty"`
+}
+
+type DataRangeResponse struct {
+	RecordType    *string    `json:"recordType,omitempty"`
+	RecordVersion *string    `json:"recordVersion,omitempty"`
+	UserID        *string    `json:"userId,omitempty"`
+	Calibrations  *DataRange `json:"calibrations,omitempty"`
+	Egvs          *DataRange `json:"egvs,omitempty"`
+	Events        *DataRange `json:"events,omitempty"`
+}
+
+func ParseDataRangeResponse(parser structure.ObjectParser) *DataRangeResponse {
+	if !parser.Exists() {
+		return nil
+	}
+	datum := NewDataRangeResponse()
+	parser.Parse(datum)
+	return datum
+}
+
+func NewDataRangeResponse() *DataRangeResponse {
+	return &DataRangeResponse{}
+}
+
+func (d *DataRangeResponse) Parse(parser structure.ObjectParser) {
+	d.UserID = parser.String("userId")
+	d.RecordType = parser.String("recordType")
+	d.RecordVersion = parser.String("recordVersion")
+	d.Calibrations = ParseDataRange(parser.WithReferenceObjectParser("calibrations"))
+	d.Egvs = ParseDataRange(parser.WithReferenceObjectParser("egvs"))
+	d.Events = ParseDataRange(parser.WithReferenceObjectParser("events"))
+}
+
+func (d *DataRangeResponse) GetOldestStartDate() time.Time {
+	oldest := time.Time{}
+	if d.Calibrations.Start != nil {
+		oldest = d.Calibrations.Start.DisplayTime.Time
+	}
+	if d.Events.Start != nil {
+		if d.Events.Start.DisplayTime.Time.Before(oldest) {
```

Comment:
I think this logic will fail if `d.Calibrations.Start` was `nil`, which leaves `oldest` still zero. In that case, no `d.Events.Start.DisplayTime` could be *before* `oldest`.

Reply:
have updated so that "oldest" begins at "now" and is updated as we get better info